src/vm/assembly.cpp

```diff
@@ -1780,6 +1780,36 @@ INT32 Assembly::ExecuteMainMethod(PTRARRAYREF *stringArgs, BOOL waitForOtherThre
     INT32   iRetVal = 0;
 
     BEGIN_ENTRYPOINT_THROWS;
+    {
+        GCX_PREEMP();
+
+        mdToken tkEntryPoint = m_pManifestFile->GetEntryPointToken();
+        // <TODO>@TODO: What if the entrypoint is in another file of the assembly?</TODO>
+        ReleaseHolder<IMDInternalImport> scope(m_pManifestFile->GetMDImportWithRef());
+        // In theory, we should have a valid executable image and scope should never be NULL, but we've been  
+        // getting Watson failures for AVs here due to ISVs modifying image headers and some new OS loader 
+        // checks (see Dev10# 718530 and Windows 7# 615596)
+        if (scope == NULL)
+        {
+            ThrowHR(COR_E_BADIMAGEFORMAT);
+        }
+
+#ifdef FEATURE_COMINTEROP
```

Comment:
Shouldn't we be setting the apartment state, regardless of whether `FEATURE_COMINTEROP` is defined? That way `Thread.GetApartmentState` returns the correct value?

Reply:
Removing this if leads to failure of tests on Unix systems. I still need to debug the reason for that

Reply:
As I have said in the other comment, the dummy apartment state on Unix is handled in C# . It does not even go to VM. This ifdef is fine. Appartments are Windows-specific COM-specific feature.

Reply:
I understand that the actual behavior the state controls is Windows (or more specifically COM) specific.

I was just worried that Unix would not get set apprioately, which could cause potential issues in applications that do a RequiresStaThread, or equivalent check.

If Unix is already handling this in the managed code path, then it is probably fine

Reply:
This is behavior on Unix: https://github.com/dotnet/corefx/blob/master/src/System.Threading.Thread/src/System/Threading/Thread.Unix.cs